packages/core/src/ckb/epoch.ts

```diff
@@ -0,0 +1,296 @@
+   */
+  normalized(): Epoch {
+    let { number, index, length } = this;
+
+    // Normalize negative index values by borrowing from the whole number.
+    if (index < Zero) {
+      // Calculate how many whole units to borrow.
+      const n = (-index + length - 1n) / length;
+      number -= n;
+      index += length * n;
+    }
+
+    // Reduce the fraction (index / length) to its simplest form using the greatest common divisor.
+    const g = gcd(index, length);
+    index /= g;
+    length /= g;
+
+    // Add any whole number overflow from the fraction.
+    number += index / length;
+
+    // Calculate the leftover index after accounting for the whole number part from the fraction.
+    index %= length;
+
+    return new Epoch(number, index, length);
+  }
+
+  /**
+   * Add another epoch to this one.
+   *
+   * If denominators differ, the method aligns to a common denominator before
+   * adding the fractional numerators, then returns a normalized Epoch.
+   *
+   * @param other - EpochLike to add.
+   * @returns New Epoch equal to this + other.
+   */
+  add(other: EpochLike): Epoch {
+    const other_ = Epoch.from(other);
+
+    // Sum the whole number parts.
+    const number = this.number + other_.number;
+    let index: Num;
+    let length: Num;
+
+    // If the epochs have different denominators (lengths), align them to a common denominator.
+    if (this.length !== other_.length) {
+      index = other_.index * this.length + this.index * other_.length;
+      length = this.length * other_.length;
+    } else {
+      // If denominators are equal, simply add the indices.
+      index = this.index + other_.index;
+      length = this.length;
+    }
+
+    return new Epoch(number, index, length).normalized();
+  }
+
+  /**
+   * Subtract an epoch from this epoch.
+   *
+   * @param other - EpochLike to subtract.
+   * @returns New Epoch equal to this - other.
+   */
+  sub(other: EpochLike): Epoch {
+    const { number, index, length } = Epoch.from(other);
+    return this.add(new Epoch(-number, -index, length));
+  }
+
+  /**
+   * Convert this epoch to an estimated Unix timestamp in milliseconds using as reference the block header.
+   *
+   * @param reference - ClientBlockHeader providing a reference epoch and timestamp.
+   * @returns Unix timestamp in milliseconds as bigint.
+   */
+  toUnix(reference: ClientBlockHeader): bigint {
+    // Calculate the difference between the provided epoch and the reference epoch.
+    const { number, index, length } = this.sub(reference.epoch);
+
+    return (
+      reference.timestamp +
+      epochInMilliseconds * number +
+      (epochInMilliseconds * index) / length
+    );
+  }
+}
+
+/**
+ * A constant representing the epoch duration in milliseconds.
+ *
+ * Calculated as 4 hours in milliseconds:
+ * 4 hours * 60 minutes per hour * 60 seconds per minute * 1000 milliseconds per second.
+ */
+const epochInMilliseconds = numFrom(14400000); // (Number.isSafeInteger(14400000) === true)
```

Comment:
Global variable names should be in UPPER_CASE, like `EPOCH_IN_MILLISECONDS`.
I suggest `numFrom(4 * 60 * 60 * 1000)` will be more readable and more maintainable.